### PR TITLE
Ability to read private catalogs

### PR DESCRIPTION
**Pull request by marksmall**
This PR covers a few tickets:

- [Add Private Catalog](https://telespazio-uk.atlassian.net/browse/EODHP-811)
- [Add Copy to clipboard](https://telespazio-uk.atlassian.net/browse/EODHP-661)
- [Add Copy to clipboard](https://telespazio-uk.atlassian.net/browse/EODHP-663)

The last 2 are effectively duplicates of each other, they were solved independently, but this now re-uses the more generic solution, developed for EODHP-663 and applied to EODHP-661.

### CODE REVIEW

**Comment by marksmall** [comment]
This whole solution still needs us to be able to configure the frontend to use the right backend, we can no longer point `DEV` to `TEST`, since we are required to forward the user to the sign in page for the current domain, not `TEST`.

This also adds a CORS problem in that requests for collections now have `credentials: 'include'`, causing this, if you comment out this line, local development will work as normal. I've used browser exension `CORS Unblock` to get around this as it is just a local development issue.

I am working on the `web-presence` API endpoint to get the right API endpoint from, instead of baking it in a Build time, as is currently done.

I'm off between 23rd to 30th, which is why I'm getting this in now, I can maybe take a look at responses, if I get bored on A/L....

**Comment by james-hinton** [comment]
The toolbox size still overflows when looking at the items (after selecting an AOI)

**Comment by marksmall** [comment]
@james-hinton I see that now, but that is not what this PR is for, that was handled in another ticket, now closed as what you found wasn't noticed. I agree it should be fixed, but not as part of this PR, it is already covering 3 separate tickets.

I would advise, unless any issues found with the 3 tickets mentioned, then this PR be approved and the appropriate ticket be re-opened, or created anew.

**Comment by james-hinton** [comment]
> @james-hinton I see that now, but that is not what this PR is for, that was handled in another ticket, now closed as what you found wasn't noticed. I agree it should be fixed, but not as part of this PR, it is already covering 3 separate tickets.
> 
> I would advise, unless any issues found with the 3 tickets mentioned, then this PR be approved and the appropriate ticket be re-opened, or created anew.

I have not completed the review as yet. Wasn't expecting you to have seen this.

**Comment by marksmall** [comment]
No worries, I was doing some learning and saw the response was 6 hours old, so figured you had finished, me jumping the gun is all....

File: src/pages/MapViewer/components/Toolbox/components/ToolboxRow/index.tsx [hunk 1]
```diff
@@ -4,10 +4,10 @@ import { ToolboxItemProps } from './types.js';
 
 import './styles.scss';
 
-const ToolboxRow = ({ thumbnail, title, dataPoints, onClick }: ToolboxItemProps) => {
+const ToolboxRow = ({ thumbnail, title, dataPoints, onClick, children }: ToolboxItemProps) => {
   return (
-    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
-    <div className="toolbox-row" onMouseUp={onClick}>
+    // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
```
Comment (james-hinton) [inline src/pages/MapViewer/components/Toolbox/components/ToolboxRow/index.tsx:9]: Should we not be changing these as we go, or is it intended to be its own separate ticket?  As I see a lot of eslint disabling going on throughout and cautious it's going to build up.

File: src/services/stac/index.tsx [hunk 2]
```diff
@@ -16,21 +18,30 @@ import { StacCollectionsResponse } from './types';
...
 ): Promise<Collection[]> => {
-  const url = `${import.meta.env.VITE_STAC_ENDPOINT}/collections?limit=${limit}&q=${searchQuery}`;
+  const url = privateCatalog
+    ? `${import.meta.env.VITE_STAC_ENDPOINT}/catalogs/${privateCatalog}/collections?limit=${limit}&q=${searchQuery}`
+    : `${import.meta.env.VITE_STAC_ENDPOINT}/collections?limit=${limit}&q=${searchQuery}`;
+
   try {
     const response = await fetch(url, {
       method: 'GET',
+      credentials: 'include',
```
Comment (james-hinton) [inline src/services/stac/index.tsx:32]: Is there a better way that we can handle this? i.e. checking if we are running on localhost?

File: src/components/clipboard/ClipboardButton.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+export const ClipboardButton = ({ text }: ClipboardButtonProps) => {
+  const [isCopied, setIsCopied] = useState(false);
+
+  const handleCopy = async (event) => {
+    event.stopPropagation();
+
+    copyTextToClipboard(text)
+      .then(() => {
```
Comment (james-hinton) [inline src/components/clipboard/ClipboardButton.tsx:29]: Was there a particular reason to use async and `.then()` rather than `await`?

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.scss [hunk 4]
```diff
@@ -0,0 +1,12 @@
+.asset-row {
+  width: 100%;
+  display: grid;
+  grid-template-columns: 140px 1fr auto;
+  grid-gap: 0.5rem;
+  padding: 0.5rem;
+
+  & .label {
```
Comment (james-hinton) [inline src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.scss:8]: Could omit the `&`

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import './AssetsPanel.scss';
+
+export const AssetsPanel = () => {
+  const {
+    state: { selectedCollectionItem },
+    actions: { setActivePage },
+  } = useToolbox();
+
+  // Reduce the object of objects down to an array of key/value objects.
+  const assets = Object.keys(selectedCollectionItem.assets).reduce((acc, key) => {
```
Comment (james-hinton) [inline src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx:13]: This can be typed: https://github.com/EO-DataHub/eodhp-resource-catalogue-ui/blob/main/src/typings/common.ts#L30

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    return acc;
+  }, []);
+
+  return (
+    <div>
+      <div className="toolbox__header">
+        <button
+          className="button-link"
+          onClick={() => {
```
Comment (james-hinton) [inline src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx:24]: I've seen that you would typically shorten this to:
``onClick={() => setActivePage('items')}``

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            setActivePage('items');
+          }}
+        >
+          <span>&lt; Return to Items</span>
+        </button>
+      </div>
+
+      <ul>
+        {assets && assets.length > 0 ? (
+          assets.map((asset) => (
```
Comment (james-hinton) [inline src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx:34]: Typed as mentioned above with the `StacAsset`

File: src/pages/MapViewer/components/Toolbox/components/ToolboxRow/index.tsx (hunk 1 above)
Comment (marksmall) [reply src/pages/MapViewer/components/Toolbox/components/ToolboxRow/index.tsx:9]: To fix this is means a bit refactoring, it means replacing any event handlers on divs,span etc with a button. The button will most likely need to be styled so it doesn't look like a button though, so there is enough going on that in my opinion, there should be a ticket just for this.

From now on though, we should not be adding event handlers to static elements, so it shouldn't build up from what we currently have.

File: src/services/stac/index.tsx (hunk 2 above)
Comment (marksmall) [reply src/services/stac/index.tsx:32]: I don't know, I'm not a fan of adding mode checks as they are easy to miss, or get wrong, but if we were to go down that road, I guess it would be something like:

```JavaScript
...(import.meta.env.PROD && {credentials : 'include'})
```

I've not proven the above works, but I don't see why it wouldn't. I would probably prefer that APIs in `DEV` environment accept requests from `localhost`, than this though, I personally find that cleaner e.g. in Django `settings/development.py`

```Python
ALLOWED_HOSTS = ["*"]
```
The CORS error is caused by the server, not the client, so I feel it is a server issue, not a client one, but I'm sure others may disagree with that opinion.

File: src/components/clipboard/ClipboardButton.tsx (hunk 3 above)
Comment (marksmall) [reply src/components/clipboard/ClipboardButton.tsx:29]: When I `await`ed the `copyTextToClipboard` function I got no response and using the `then` was how the example I used did it. I did this over a week ago now and I can't remember what source I used. 

Honestly, I didn't give it any real thought, since the `await` returned nothing and I only wanted to do the `then` bit if successful, I just used this method. I could have used a `try/catch` but I tend to find them overly verbose and ugly for async stuff, this just seemed that little bit neater to me.

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.scss (hunk 4 above)
Comment (marksmall) [reply src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.scss:8]: your right, I will remove this. Been a few years since I used SASS, but rusty....

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx (hunk 5 above)
Comment (marksmall) [reply src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx:13]: Not really, or at least not as you suggest. The `assets` variable is not a `StacAsset[]`, but a simple array of objects of `key/href` e.g.

```JavaScript
[
  { cloud: 'http://....' },
  ...
  { valid_pixels: 'http://....' },
]
```
I am find with creating a local `type`:

```JavaScript
type Asset = {
  key: string;
  value: string;
}
```
That would satisfy typing of the reduced array, but since the reduce is using only the `keys` from the assets and not the assets from the CollectionItem themselves, what you suggest wouldn't work.

the StacAsset is an object and I only care about the key name and the href value, so I tend to find it easier to reduce this object down to a simple array of new objects, in the shape that is easy to iterate, to create the label/input/button structure. I could not do this and instead of the `assets.map` later, do a more complicated extraction of the keys and href value there. The code would be very similar but I feel that makes the code less readable.

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx (hunk 6 above)
Comment (marksmall) [reply src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx:24]: good catch, I'll simplify. I tend to write it out like this first then remove the curly braces once ready to commit, obviously I missed this out here.

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx (hunk 7 above)
Comment (marksmall) [reply src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx:34]: See comment above on this. At best, this would be a new type as I stated, not a `StacAsset`.

File: src/services/stac/index.tsx (hunk 2 above)
Comment (james-hinton) [reply src/services/stac/index.tsx:32]: ^^
This refers to the CORS that comes through the OPA. I tried to resolve it a while ago and it seems unresolvable.

File: src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx (hunk 5 above)
Comment (james-hinton) [reply src/pages/MapViewer/components/Toolbox/components/item-assets/AssetsPanel.tsx]: Okay good point

**Review by james-hinton — APPROVED** [review]
